_posts/2020-04-07-how-to-start-with-opentelemetry-in-nodejs.markdown

```diff
@@ -0,0 +1,300 @@
+"@opentelemetry/plugin-dns": "^0.5.0",
+"@opentelemetry/plugin-mongodb": "^0.5.0",
+"@opentelemetry/tracing": "^0.5.0",
+"@opentelemetry/plugin-express": "^0.5.0",
+```
+
+I created a `./tracer.js` file that initialize the tracer, I have added inline
+documentation to explain the crucial part of it:
+
+```
+js'use strict';
+
+const opentelemetry = require('@opentelemetry/api');
+const { NodeTracerProvider } = require('@opentelemetry/node');
+const { SimpleSpanProcessor } = require('@opentelemetry/tracing');
+// I am using Jaeger as exporter
+const { JaegerExporter } = require('@opentelemetry/exporter-jaeger');
+
+// This is not mandatory, by default httptrace propagation is used
+// but it is not well supported by the PHP ecosystem and I have
+// a PHP service to instrument. I discovered B3 is supported
+// form all the languages I where intrumenting
+const { B3Propagator } = require('@opentelemetry/core');
+
+module.exports = (serviceName, jaegerHost, logger) => {
+  // A lot of those plugins are automatically loaded when you install them
+  // So if you do not use express for example you do not have to enable all
+  // those plugins manually. But Express is not auto enabled so I had to add them
+  // all
+  const provider = new NodeTracerProvider({
+    plugins: {
+      dns: {
+        enabled: true,
+        path: '@opentelemetry/plugin-dns',
+      },
+      mongodb: {
+        enabled: true,
+        path: '@opentelemetry/plugin-mongodb',
+      },
+      http: {
+        enabled: true,
+        path: '@opentelemetry/plugin-http',
+      },
+      express: {
+        enabled: true,
+        path: '@opentelemetry/plugin-express',
+      },
+    }
+  });
+
+  // Here is where I configured the exporter, setting the service name
+  // and the jaeger host. The logger is helpful to track errors from the
+  // exporter itself
+  let exporter = new JaegerExporter({
+    logger: logger,
+    serviceName: serviceName,
+    host: jaegerHost
+  });
+
+  provider.addSpanProcessor(new SimpleSpanProcessor(exporter));
+  provider.register({
+    propagator: new B3Propagator(),
+  });
+  // Set the global tracer so you can retrieve it from everywhere else in the
+  // app
+  return opentelemetry.trace.getTracer("discount");
+};
+```
+
+You will be thinking, that's too easy! You are right, the nature of NodeJS
+makes tracing very code agnostic. With this configuration you get a lot "for
+free".
+
+You get a bunch of spans for every http request that ExpressJS serves, plus a
+span for every MongoDB query. All of them with useful information like the
+status code, path, user agents, query statements and so on.
+
+We have to include it in our `./server.js` the entrypoint for our nodejs
+application:
+
+```js
+'use strict';
+
+const url = process.env.DISCOUNT_MONGODB_URL || 'mongodb://discountdb:27017';
+const jaegerHost = process.env.JAEGER_HOST || 'jaeger';
+
+const logger = require('pino')()
+
+// Import and initialize the tracer
```

Comment:
Important to note that it generally need to be the first module to be required in your application otherwise some automatic instrumentation might not work